Support map in pre-dispatch functionalization

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121444

When we enter map_autograd, we try to trace through fwd/bwd of a map operator that is wrapped in ctx.functionalize wrapper. This forces us to go through PreDispatch functionalization again (only the python part). As a result, it revealed our previous bug where pre-dispatch mode handling doesn't actually manage the local dispatch key set. (If there is no active mode, we need to turn off PreDispatch key). This PR fixes that. Also I shuffled some APIs around so that there is less code duplication as the setting/unsetting logic is quite hard to get it right.

tugsbayasgalan committed:
Support map in pre-dispatch functionalization

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"


When we enter map_autograd, we try to trace through fwd/bwd of a map operator that is wrapped in ctx.functionalize wrapper. This forces us to go through PreDispatch functionalization again (only the python part). As a result, it revealed our previous bug where pre-dispatch mode handling doesn't actually manage the local dispatch key set. (If there is no active mode, we need to turn off PreDispatch key). This PR fixes that. Also I shuffled some APIs around so that there is less code duplication as the setting/unsetting logic is quite hard to get it right. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"


When we enter map_autograd, we try to trace through fwd/bwd of a map operator that is wrapped in ctx.functionalize wrapper. This forces us to go through PreDispatch functionalization again (only the python part). As a result, it revealed our previous bug where pre-dispatch mode handling doesn't actually manage the local dispatch key set. (If there is no active mode, we need to turn off PreDispatch key). This PR fixes that. Also I shuffled some APIs around so that there is less code duplication as the setting/unsetting logic is quite hard to get it right. 



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"


When we enter map_autograd, we try to trace through fwd/bwd of a map operator that is wrapped in ctx.functionalize wrapper. This forces us to go through PreDispatch functionalization again (only the python part). As a result, it revealed our previous bug where pre-dispatch mode handling doesn't actually manage the local dispatch key set. (If there is no active mode, we need to turn off PreDispatch key). This PR fixes that. Also I shuffled some APIs around so that there is less code duplication as the setting/unsetting logic is quite hard to get it right. 



[ghstack-poisoned]

bdhirsh reviewed (approved):
nice and clean :)

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"


When we enter map_autograd, we try to trace through fwd/bwd of a map operator that is wrapped in ctx.functionalize wrapper. This forces us to go through PreDispatch functionalization again (only the python part). As a result, it revealed our previous bug where pre-dispatch mode handling doesn't actually manage the local dispatch key set. (If there is no active mode, we need to turn off PreDispatch key). This PR fixes that. Also I shuffled some APIs around so that there is less code duplication as the setting/unsetting logic is quite hard to get it right. 



[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot merge -f "Unrelated failure"

clee2000 commented:
@pytorchbot revert -m "sorry windows failure seems related https://hud.pytorch.org/pytorch/pytorch/commit/079feea3379c021a330dbfac7668a5fc8fccc3bd https://github.com/pytorch/pytorch/actions/runs/8474191301/job/23220791555. PR got force merged before windows job finished" -c ignoredsignal

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"


When we enter map_autograd, we try to trace through fwd/bwd of a map operator that is wrapped in ctx.functionalize wrapper. This forces us to go through PreDispatch functionalization again (only the python part). As a result, it revealed our previous bug where pre-dispatch mode handling doesn't actually manage the local dispatch key set. (If there is no active mode, we need to turn off PreDispatch key). This PR fixes that. Also I shuffled some APIs around so that there is less code duplication as the setting/unsetting logic is quite hard to get it right. 



[ghstack-poisoned]

atalman commented:
@pytorchmergebot revert -m "New test test_aot_export_predispatch_map_1 is failing on windows" -c ignoredsignal

tugsbayasgalan committed:
Update on "Support map in pre-dispatch functionalization"


When we enter map_autograd, we try to trace through fwd/bwd of a map operator that is wrapped in ctx.functionalize wrapper. This forces us to go through PreDispatch functionalization again (only the python part). As a result, it revealed our previous bug where pre-dispatch mode handling doesn't actually manage the local dispatch key set. (If there is no active mode, we need to turn off PreDispatch key). This PR fixes that. Also I shuffled some APIs around so that there is less code duplication as the setting/unsetting logic is quite hard to get it right. 



[ghstack-poisoned]
